File: libraries/android/sdk/src/main/java/com/voidhash/sdk/schema/RuntimeSchema.kt
package com.voidhash.sdk.schema

import org.json.JSONObject

/** Google Play provider configuration of a product. */
data class RuntimeGooglePlayConfiguration(
    val productId: String,
    val basePlanId: String?,
)

/** Apple App Store provider configuration of a product. */
data class RuntimeAppleAppStoreConfiguration(
    val productId: String,
)

/** Development-provider configuration of a product, used by the mock store. */
data class RuntimeDevelopmentConfiguration(
    val productId: String,
    val price: Double,
    val currencyCode: String,
    val period: String,
    val periodCount: Int,
    val duration: String?,
    val warning: String?,
)

/** Store providers a product is configured for. */
data class RuntimeProductProviders(
    val googlePlay: RuntimeGooglePlayConfiguration? = null,
    val appleAppStore: RuntimeAppleAppStoreConfiguration? = null,
    val development: RuntimeDevelopmentConfiguration? = null,
)

/** A product as defined in the project schema. */
data class RuntimeProductDefinition(
    val slug: String,
    val type: String,
    val name: String,
    val providers: RuntimeProductProviders,
)

/** A paywall location as defined in the project schema. */
data class RuntimeLocationDefinition(
    val slug: String,
    val name: String,
    val description: String?,
)

/** The project schema as fetched from `/api/v1/sdk/schema`, keyed by slug. */
data class RuntimeSchema(
    val version: String,
    val products: Map<String, RuntimeProductDefinition>,
    val locations: Map<String, RuntimeLocationDefinition>,
) {
    companion object {
        /**
         * The schema a client has when the backend has never been reachable.
         *
         * Store operations answer from it rather than failing: an app that asks for its
         * products on a cold offline launch gets an empty list, which it can render, instead
         * of an exception it has to catch.
         */
        val EMPTY: RuntimeSchema = RuntimeSchema("", emptyMap(), emptyMap())

        /** Parses the schema response body. */
        fun fromJson(json: JSONObject): RuntimeSchema {
            val products = json.optJSONObject("products") ?: JSONObject()
            val locations = json.optJSONObject("locations") ?: JSONObject()

            return RuntimeSchema(
                version = json.optString("version"),
                products = products.keys().asSequence().associateWith { slug ->
                    val product = products.getJSONObject(slug)
                    val providers = product.optJSONObject("configuration")
                        ?.optJSONObject("providers")
                    RuntimeProductDefinition(
                        slug = product.optString("slug", slug),
                        type = product.optString("type"),
                        name = product.optJSONObject("properties")?.optString("name").orEmpty(),
                        providers = RuntimeProductProviders(
                            googlePlay = providers?.optJSONObject("googlePlay")?.let {
                                RuntimeGooglePlayConfiguration(
                                    productId = it.optString("productId"),
                                    basePlanId = if (it.isNull("basePlanId")) {
                                        null
                                    } else {
                                        it.optString("basePlanId").ifEmpty { null }
                                    },
                                )
                            },
                            appleAppStore = providers?.optJSONObject("appleAppStore")?.let {
                                RuntimeAppleAppStoreConfiguration(it.optString("productId"))
                            },
                            development = providers?.optJSONObject("development")?.let {
                                RuntimeDevelopmentConfiguration(
                                    productId = it.optString("productId"),
                                    price = it.optDouble("price"),
                                    currencyCode = it.optString("currencyCode"),
                                    period = it.optString("period"),
                                    periodCount = it.optInt("periodCount", 1),
                                    duration = if (it.isNull("duration")) {
                                        null
                                    } else {
                                        it.optString("duration").ifEmpty { null }
                                    },
                                    warning = if (it.isNull("warning")) {
                                        null
                                    } else {
                                        it.optString("warning").ifEmpty { null }
                                    },
                                )
                            },
                        ),
                    )
                },
                locations = locations.keys().asSequence().associateWith { slug ->
                    val location = locations.getJSONObject(slug)
                    RuntimeLocationDefinition(
                        slug = location.optString("slug", slug),
                        name = location.optString("name"),
                        description = if (location.isNull("description")) {
                            null
                        } else {
                            location.optString("description").ifEmpty { null }
                        },
                    )
                },
            )
        }
    }
}

File: libraries/ios/Sources/VoidhashCore/Schema/RuntimeSchema.swift
import Foundation

/// App Store configuration of a product.
public struct RuntimeAppleAppStoreProductConfiguration: Codable, Sendable, Equatable {
    public let productId: String

    public init(productId: String) {
        self.productId = productId
    }
}

/// Google Play configuration of a product.
public struct RuntimeGooglePlayProductConfiguration: Codable, Sendable, Equatable {
    public let productId: String
    public let basePlanId: String?

    public init(productId: String, basePlanId: String? = nil) {
        self.productId = productId
        self.basePlanId = basePlanId
    }
}

/// Development-provider configuration of a product, used by the mock store.
public struct RuntimeDevelopmentProductConfiguration: Codable, Sendable, Equatable {
    public let currencyCode: String
    public let duration: String?
    public let period: String
    public let periodCount: Double
    public let price: Double
    public let priceInMinorUnits: Double
    public let productId: String
    public let warning: String?

    public init(
        currencyCode: String,
        duration: String?,
        period: String,
        periodCount: Double,
        price: Double,
        priceInMinorUnits: Double,
        productId: String,
        warning: String?
    ) {
        self.currencyCode = currencyCode
        self.duration = duration
        self.period = period
        self.periodCount = periodCount
        self.price = price
        self.priceInMinorUnits = priceInMinorUnits
        self.productId = productId
        self.warning = warning
    }
}

/// Per-provider configuration of a product.
public struct RuntimeProductProviders: Codable, Sendable, Equatable {
    public let appleAppStore: RuntimeAppleAppStoreProductConfiguration?
    public let development: RuntimeDevelopmentProductConfiguration?
    public let googlePlay: RuntimeGooglePlayProductConfiguration?

    public init(
        appleAppStore: RuntimeAppleAppStoreProductConfiguration? = nil,
        development: RuntimeDevelopmentProductConfiguration? = nil,
        googlePlay: RuntimeGooglePlayProductConfiguration? = nil
    ) {
        self.appleAppStore = appleAppStore
        self.development = development
        self.googlePlay = googlePlay
    }
}

/// Configuration block of a product definition.
public struct RuntimeProductConfiguration: Codable, Sendable, Equatable {
    public let providers: RuntimeProductProviders
    public let perks: [String: Bool]

    public init(providers: RuntimeProductProviders, perks: [String: Bool] = [:]) {
        self.providers = providers
        self.perks = perks
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        providers = try container.decode(RuntimeProductProviders.self, forKey: .providers)
        perks = try container.decodeIfPresent([String: Bool].self, forKey: .perks) ?? [:]
    }
}

/// Human readable properties of a product definition.
public struct RuntimeProductProperties: Codable, Sendable, Equatable {
    public let name: String

    public init(name: String) {
        self.name = name
    }
}

/// A product as configured in the dashboard.
public struct RuntimeProductDefinition: Codable, Sendable, Equatable {
    public let id: String?
    public let duration: String?
    public let slug: String
    public let type: String
    public let properties: RuntimeProductProperties
    public let configuration: RuntimeProductConfiguration

    public init(
        id: String? = nil,
        duration: String? = nil,
        slug: String,
        type: String,
        properties: RuntimeProductProperties,
        configuration: RuntimeProductConfiguration
    ) {
        self.id = id
        self.duration = duration
        self.slug = slug
        self.type = type
        self.properties = properties
        self.configuration = configuration
    }
}

/// A paywall location as configured in the dashboard.
public struct RuntimePaywallLocationDefinition: Codable, Sendable, Equatable {
    public let slug: String
    public let name: String
    public let description: String?

    public init(slug: String, name: String, description: String? = nil) {
        self.slug = slug
        self.name = name
        self.description = description
    }
}

/// The runtime schema fetched from `/api/v1/sdk/schema`, keyed by slug.
///
/// Mirrors `src/core/schema/runtime.ts`; `version` is the server-side schema hash.
public struct RuntimeSchema: Codable, Sendable, Equatable {
    public let version: String
    public let products: [String: RuntimeProductDefinition]
    public let locations: [String: RuntimePaywallLocationDefinition]

    public init(
        version: String,
        products: [String: RuntimeProductDefinition] = [:],
        locations: [String: RuntimePaywallLocationDefinition] = [:]
    ) {
        self.version = version
        self.products = products
        self.locations = locations
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        version = try container.decode(String.self, forKey: .version)
        products =
            try container.decodeIfPresent(
                [String: RuntimeProductDefinition].self, forKey: .products) ?? [:]
        locations =
            try container.decodeIfPresent(
                [String: RuntimePaywallLocationDefinition].self, forKey: .locations) ?? [:]
    }

    /// The App Store product id configured for `slug`, if any.
    public func appleProductId(forSlug slug: String) -> String? {
        return products[slug]?.configuration.providers.appleAppStore?.productId
    }
}
